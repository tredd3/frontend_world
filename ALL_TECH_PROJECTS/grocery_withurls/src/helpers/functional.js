export const pipe = (...args) => x => args.reduce((acc, fn) => fn(acc), x);

export const range = length => [...Array(length).keys()];

export const pad = (length, padCharacter) => value => {
  if (String(value).length >= length) return String(value);
  return range(length - String(value).length).map(() => padCharacter).join('') + String(value);
};

const make2digit = pad(2, '0');

export const formatDate = dateLike => {
  const date = new Date(dateLike);
  return `${date.getFullYear()}-${make2digit(date.getMonth() + 1)}-${make2digit(date.getDate())}`;
};

export const objectToQueryString = obj => Object.entries(obj)
  .reduce((acc, [key, value]) => [...acc, `${encodeURIComponent(key)}=${encodeURIComponent(Array.isArray(value)
    ? value.join(',') : value)}`], []).join('&');

export const queryStringToObject = qs => {
  if (!qs) return {};
  return qs.split('&')
    .map(aQsPart => aQsPart.split('=').map(decodeURIComponent))
    .reduce((acc, [name, value]) => {
      const convertedValue = value.includes(',') ? value.split(',') : [value];
      return { ...acc, [name]: convertedValue };
    }, {});
};
