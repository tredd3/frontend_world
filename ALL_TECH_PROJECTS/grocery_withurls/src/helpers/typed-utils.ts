export const flatMap = <T, U>(arr: T[], fn: ((x: T) => U[])): U[] => (
  arr.reduce(
    (acc: U[], x: T) => [...acc, ...fn(x)],
    [] as U[]
  )
);
