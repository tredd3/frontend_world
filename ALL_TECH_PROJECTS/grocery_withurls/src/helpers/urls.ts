export const accountUrl = '/account';
export const cartUrl = '/cart';
export const notificationsUrl = '/notifications';
export const productUrl = (skuId: number | string) => `/products/${skuId}`;
export const productSimilarItemsUrl = (skuId: number | string) => `/products/${skuId}/similar-items`;
export const orderPlacedUrl = (orderId: number | string) => `/order-placed/${orderId}`;
export const ordersUrl = '/orders';
export const orderSearchUrl = '/orders/search';
export const orderDetailsUrl = (orderId: number | string) => `/orders/${orderId}`;
export const wishlistUrl = '/wishlist';
export const categoriesUrl = '/categories';
export const categoryUrl = (categoryId: number | string) => `/categories/${categoryId}`;
export const subCategoryUrl = (categoryId: number | string, subCategoryId: number | string) => `/categories/${categoryId}/${subCategoryId}`;
export const searchUrl = '/search';
export const searchResultsUrl = (searchText: string) => `/search/${searchText}`;
export const collectionUrl = (collectionId: number | string, type?: number | string, subtype?: number | string) => {
  let queryParams;
  if (type) {
    queryParams = new URLSearchParams();
    queryParams.append('type', String(type));

    if (subtype) {
      queryParams.append('subtype', String(subtype));
    }
  }
  return `/collections/${collectionId}${queryParams ? `?${queryParams.toString()}` : ''}`;
};
