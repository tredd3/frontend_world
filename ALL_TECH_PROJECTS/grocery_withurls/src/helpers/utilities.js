import store from '../store';
import { queryStringToObject } from './functional';

const LIFTED_SNACKBAR_ROUTES = ['/', '/static/jiokart/index.html',
  '/index.html', '/static/jiomart/index.html', '/static/jiomart/', '/static/jiomart/cart/addresses'];

export function getRan(n = 5) {
  let rand = Math.round(Math.random() * 10 ** n);
  if (rand.toString().length < n) {
    const restLength = n - rand.toString().length;
    const zeroes = (() => {
      let zero = '';
      for (let i = 0; i < restLength; i += 1) {
        zero += '0';
      }
      return zero;
    })();
    rand += zeroes;
  }
  return rand.toString();
}

export const numberOfDigits = num => Math.floor(Math.log10(num) + 1);

// Returns a function, that, as long as it continues to be invoked, will not
// be triggered. The function will be called after it stops being called for
// N milliseconds. If `immediate` is passed, trigger the function on the
// leading edge, instead of the trailing.
export function debounce(func, wait, immediate) {
  let timeout;
  return function debounceInner() {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const context = this;
    // eslint-disable-next-line prefer-rest-params
    const args = arguments;
    const later = function later() {
      timeout = null;
      if (!immediate) func.apply(context, args);
    };
    const callNow = immediate && !timeout;
    clearTimeout(timeout);
    timeout = setTimeout(later, wait);
    if (callNow) func.apply(context, args);
  };
}

function getDate(dateObj) {
  const date = (`0${dateObj.getDate()}`).slice(-2);
  const month = (`0${dateObj.getMonth() + 1}`).slice(-2);
  const year = dateObj.getFullYear();
  return `${year}-${month}-${date}`;
}

export function getDateFromEpoch(ts) {
  const utcSeconds = Math.floor(Number(ts) / 1000);
  const d = new Date(0); // The 0 there is the key, which sets the date to the epoch
  d.setUTCSeconds(utcSeconds);
  return getDate(d);
}

export function formatTime(time) {
  const timeArray = time.split(':');
  let hour = timeArray[0];
  const min = timeArray[1];
  let formattedTime = '';
  if (hour > 12) {
    hour -= 12;
    formattedTime = `${hour}:${min} pm`;
  } else {
    formattedTime = `${hour}:${min} am`;
  }
  return formattedTime;
}

export function formatDate(date, monthTitleCase = false) {
  const od = date.split(' ')[0].split('-');
  const day = od[2];
  const month = Number(od[1]) - 1;
  const year = od[0];
  const months = [
    'JAN',
    'FEB',
    'MAR',
    'APR',
    'MAY',
    'JUN',
    'JUL',
    'AUG',
    'SEP',
    'OCT',
    'NOV',
    'DEC'
  ];
  if (monthTitleCase) {
    return `${day} ${months[month].replace(/\w\S*/g, txt => txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase())} ${year}`;
  }
  return `${day} ${months[month]} ${year}`;
}

export function decodeDate(date, gap) {
  const day = date.getDate();
  const month = date.getMonth();
  const year = date.getFullYear();
  const map = new Map([
    [0, 31],
    [1, 28],
    [2, 31],
    [3, 30],
    [4, 31],
    [5, 30],
    [6, 31],
    [7, 31],
    [8, 30],
    [9, 31],
    [10, 30],
    [11, 31]
  ]);
  const months = [
    'JAN',
    'FEB',
    'MAR',
    'APR',
    'MAY',
    'JUN',
    'JUL',
    'AUG',
    'SEP',
    'OCT',
    'NOV',
    'DEC'
  ];
  if (gap === '30days') {
    let customDay; let customMonth; let
      customYear = year;
    if (day >= 30) {
      customDay = '01';
    } else {
      customMonth = month - 1;
      if (customMonth < 0) {
        customYear = year - 1;
        customMonth = 11;
      }
      customDay = map.get(customMonth) + day - 30;
    }

    return `${customDay}-${months[customMonth]}-${customYear}`;
  } if (gap === '6months') {
    let customMonth; let
      customYear = year;
    if (month < 6) {
      customMonth = month + 7;
      customYear = year - 1;
    } else {
      customMonth = month - 6;
    }
    return `${day}-${months[customMonth]}-${customYear}`;
  }
  return `${day}-${months[month]}-${year}`;
}

// cart quantity helper
export function getProductQty(productSkuId) {
  const { prdQtyObj } = store.getState().cart;
  const prdIdArr = Object.keys(prdQtyObj);
  return prdIdArr.includes(productSkuId.toString()) ? prdQtyObj[productSkuId] : 0;
}

export function getMobileOperatingSystem() {
  const userAgent = navigator.userAgent || navigator.vendor || window.opera;

  // Windows Phone must come first because its UA also contains "Android"
  if (/windows phone/i.test(userAgent)) {
    return 1;// "Windows Phone"
  }

  if (/android/i.test(userAgent)) {
    return 2;// "Android"
  }

  // iOS detection from: http://stackoverflow.com/a/9039885/177710
  if (/iPad|iPhone|iPod/.test(userAgent) && !window.MSStream) {
    return 3;// "iOS"
  }

  return 4; // unknown
}

export const getRedirectUrl = querySearch => {
  const queryParam = new URLSearchParams(querySearch);

  if (queryParam.get('from') === 'checkout') {
    return `/account/select-address${querySearch}`;
  }
  return `/account/addresses${querySearch}`;
};

export const getUrlParams = param => {
  const search = decodeURIComponent(window.location.search);
  const searchParam = new URLSearchParams(search);
  return searchParam.get(param);
};

export const updateSelectedFilters = (selectedFilters, filter, canSelectMultipleValues) => {
  const { id, value, checked } = filter;
  const updatedFilters = { ...selectedFilters };

  if (!checked) {
    if (updatedFilters[id]) {
      updatedFilters[id] = updatedFilters[id].filter(filterValue => filterValue !== value);

      if (!updatedFilters[id].length) {
        delete updatedFilters[id];
      }
    }
  } else if (canSelectMultipleValues) {
    if (!updatedFilters[id]) {
      updatedFilters[id] = [];
    }

    updatedFilters[id].push(value);
  } else {
    updatedFilters[id] = [value];
  }

  return updatedFilters;
};

export const parseQSForFiltersAndSort = (baseFilters, queryString, excludeKeysFromQS = []) => {
  const filter = { ...baseFilters, '@type': 'Filter' };
  const { sort: sortObj, ...queryFilterObj } = queryStringToObject(queryString);
  excludeKeysFromQS.forEach(keyName => delete queryFilterObj[keyName]);

  Object.keys(queryFilterObj).forEach(keyName => {
    filter[keyName] = queryFilterObj[keyName];
  });

  let sort = {
    _score: 'desc',
    isFc: 'desc',
    isUKA: 'desc'
  };

  try {
    if (sortObj && sortObj[0]) {
      sort = { ...sort, ...JSON.parse(sortObj[0]) };
    }
  } catch (e) {
    // Do nothing if sort object isn't parseable
  }

  return { filter, sort };
};

export const parseQSForAppliedFiltersCount = (filters, queryString) => {
  let count = 0;
  const { sort, ...queryObj } = queryStringToObject(queryString);

  filters.forEach(filter => {
    count = queryObj[filter.id] ? count + queryObj[filter.id].length : count;
  });

  if (sort && sort[0]) {
    count += 1;
  }

  return count;
};

export const truncateWithEllipsis = text => ((text && text !== null && text.length > 10)
  ? (`${(text).substring(0, 9)}â€¦`)
  : text);

export const shouldLiftSnackbar = () => (LIFTED_SNACKBAR_ROUTES.includes(window.location.pathname) && getMobileOperatingSystem() !== 3);

const liftedRoutes = ['/', '/static/jiokart/index.html',
  '/index.html', '/static/jiomart/index.html', '/static/jiomart/', '/static/jiomart/cart/addresses'];

export const getBottomOffset = () => (liftedRoutes.includes(window.location.pathname)
&& getMobileOperatingSystem() !== 3 ? 55 : 0);

export const trimLowerCase = value => (value ? (value.replace(/ /g, '').toLowerCase() || '') : '');

export const setFromMyJioSearch = () => {
  if (getUrlParams('myjiosearch')) {
    sessionStorage.setItem('myjiosearch', true);
  }
};
export const isFromMyJioSearch = () => sessionStorage.getItem('myjiosearch');

export const removeFromMyJioSearch = () => sessionStorage.removeItem('myjiosearch');

