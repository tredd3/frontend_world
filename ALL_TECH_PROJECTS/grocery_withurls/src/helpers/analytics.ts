import { getMobileOperatingSystem } from './utilities';
import { getUser } from '../services/user';
import { getCart, PaymentMethodId, paymentLabelById } from '../services/cart';

import { CheckoutData } from '../services/checkout';
import { CartPivot } from '../types';
import { flatMap } from './typed-utils';

let purchaseJourney: string;

export const setPurchaseJourney = (val: string) => {
  purchaseJourney = val;
};

export const getPurchaseJourney = () => purchaseJourney;

declare global {
  // eslint-disable-next-line @typescript-eslint/consistent-type-definitions
  interface Window {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    digitalData: Record<string, any>;
    _satellite: {
      track: (event: string) => void;
    };
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const track = async (event: string, data: Record<string, any>) => {
  if (typeof window === 'undefined') return;
  const { userId } = await getUser();
  const userOS = getMobileOperatingSystem();
  const platform = userOS === 2 ? 'Android App' : userOS === 3 ? 'iOS App' : 'Web';

  const digitalData = {
    link: {
      ...data.link
    },
    internal: {
      ...data.internal
    },
    page: {
      pageInfo: {
        appName: 'JioMart Web'
      },
      ...data.page
    },
    user: {
      platform,
      loginStatus: 'Logged in',
      martId: userId,
      jioCenter: '',
      ...data.user
    },
    product: [
      ...(data.product ? data.product : [])
    ],
    interPromotion: {
      ...data.interPromotion
    },
    ...(data.product ? data.filterSection : {}),
    ...(data.purchase ? data.purchase : {})
  };

  window.digitalData = {
    ...window.digitalData,
    ...digitalData
  };

  // eslint-disable-next-line no-underscore-dangle
  if (window._satellite) window._satellite.track(event);
};

type LinkPositions = 'Top' | 'Middle' | 'Bottom';

export const trackLink = (linkName: string, linkType: string, linkPosition: LinkPositions) => (
  track('linkTracking', { link: { linkName, linkType, linkPosition } })
);

export const trackPage = (pageName: string) => (
  track('pageLoad', { page: { pageName } })
);

type FilterSection = {
  filterSelection: boolean;
  sortSelection: boolean;
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const trackProduct = (eventName: string, productData: Record<string, any>, filterSection?: FilterSection) => {
  const defaultFilterSection = {
    filterSelection: false,
    sortSelection: false
  };
  return track(eventName, {
    product: [productData],
    filterSection: (filterSection || defaultFilterSection)
  });
};

export const trackAPIShipments = (eventName: string, shipments: CheckoutData['Shipments'], purchase?: any) => {
  if (shipments && shipments.length) {
    track(eventName, {
      product: flatMap(shipments, shipment => shipment.Product)
        .map(product => ({
          productID: product.ProductSkuId,
          quantity: product.SelectedQuantity,
          price: product.SelectedQuantity * product.SP,
          productName: product.ProductName
        })),
      ...(purchase && { purchase })
    });
  }
};

type PurchaseArg = {
  paymentMethod: PaymentMethodId;
  paymentStatus?: 'Paid' | 'Unpaid';
  cart: CartPivot;
  failureReason?: string;
}

export const trackShipments = async (eventName: string, purchase?: PurchaseArg) => {
  const { shipments } = purchase?.cart || await getCart();
  if (!shipments || !shipments.length) return Promise.resolve();

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let additionalInfo: Record<string, any> | null = null;

  if (purchase) {
    const { paymentMethod, cart, ...rest } = purchase;
    additionalInfo = {
      paymentMethod: paymentLabelById(paymentMethod),
      couponDiscount: cart.couponDiscountAmount,
      deliveryCharges: cart.deliveryCharge,
      totalOrderValue: cart.totalPayableAmount,
      transactionId: cart.cartId,
      ...rest
    };
  }

  return track(eventName, {
    product: flatMap(shipments, shipment => shipment.products)
      .map(product => ({
        productId: product.skuId,
        quantity: product.selectedQuantity,
        price: product.selectedQuantity * product.sp,
        productName: product.name
      })),
    ...(additionalInfo && additionalInfo)
  });
};
