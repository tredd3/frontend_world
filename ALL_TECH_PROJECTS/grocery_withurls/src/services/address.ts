import promiseMemoize from 'promise-memoize';
import makeApiCall from './helpers/make-api-call';
import { apiRoutes } from './api-routes';
import { getUser, updateUser } from './user';
import { APIBaseAddress, APIAddress, AddressPivot } from '../types';
import { pivotAPIAddress } from './helpers/pivots';

export type APIAddressRequest = APIBaseAddress & {
  addressId: number;
  addedDate: string;
};

// eslint-disable-next-line camelcase, @typescript-eslint/camelcase
export const DEPRECATED_convertAddressToAPIAddress = ({ id, addedDate, ...address }: APIAddress): APIAddressRequest => (
  { addressId: id, addedDate: addedDate.toDateString(), ...address }
);

type AreaType = { pincode: number};
export const isAreaServiceable = promiseMemoize(async ({ pincode }: AreaType) => {
  try {
    await makeApiCall<AreaType, {}>(apiRoutes.checkServiceability, { body: { pincode } });
    return true;
  } catch (e) {
    return false;
  }
}, { resolve: 'json' });

export const getAddresses = promiseMemoize(async (): Promise<AddressPivot[]> => {
  const { userId } = await getUser();
  const { addresses } = await makeApiCall<{ userId: number }, {addresses: APIAddressRequest[]}>(apiRoutes.getAddress, { body: { userId } });
  return addresses.map(pivotAPIAddress);
});

export const getAddressById = async (id: number): Promise<AddressPivot | undefined> => {
  const addresses = (await getAddresses()) || [];
  return addresses.find(address => address.id === id);
};

type UpdateAddressRequest = (APIAddressRequest | APIBaseAddress) & { userId?: number };

export const updateAddress = async (address: Omit<APIBaseAddress, 'address'> | APIAddress): Promise<APIAddress> => {
  const { id } = address as APIAddress;
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { addedDate, ...remainingAddress } = address as APIAddress;
  const newAddress = await makeApiCall<UpdateAddressRequest, APIAddressRequest>(apiRoutes.saveAddress, {
    body: {
      ...(id && { addressId: id }),
      ...({ userId: (await getUser()).userId }),
      ...remainingAddress,
      address: `${remainingAddress.cityName}, ${remainingAddress.pincode}`
    }
  });

  getAddresses.clear();
  return pivotAPIAddress(newAddress);
};

export const addAddress = (address: Omit<APIBaseAddress, 'address'>) => updateAddress(address);

export const deleteAddress = async (id: number) => {
  getAddresses.clear();
  return makeApiCall<{ addressId: number }, void>(
    apiRoutes.deleteAddress, { body: { addressId: id } }
  );
};

export const setDefaultAddress = async (id: number) => {
  const address = await getAddressById(id);
  if (!address) throw new Error('An address with this id doesn\'t exist');

  await makeApiCall<{ addressId: number; isDefault: boolean }, APIAddressRequest>(apiRoutes.saveAddress, {
    body: {
      addressId: id,
      isDefault: true
    }
  });
  getAddresses.clear();

  await updateUser({
    lat: address.latitude,
    lng: address.longitude,
    pincode: address.pincode,
    storeId: address.storeId,
    storeName: address.storeName
  });

  return address;
};

export const getDefaultAddress = async () => {
  const addresses = await getAddresses();
  const defaultAddress = addresses.find(address => address.isDefault);
  if (defaultAddress) return defaultAddress;

  if (!addresses.length) return null;

  await setDefaultAddress(addresses[0].id);
  return addresses[0];
};

export const updateAddressWithKirana = async (addressId: number, { storeId, storeName }: { storeId: number; storeName: string }) => {
  const address = await getAddressById(addressId);

  if (!address) throw new Error('Couldn\'t find an address with this id');

  const updatedAddress = await updateAddress({ ...address, storeId, storeName });

  if (address.isDefault) {
    await updateUser({ storeId, storeName });
  }

  return updatedAddress;
};

export const deleteKiranaFromAddress = async (addressId: number) => updateAddressWithKirana(addressId, { storeId: 0, storeName: '' });

export const getStoreIdFromAddressId = async (addressId: number) => {
  const address = await getAddressById(addressId);

  if (!address) return null;

  return address.storeId || address.defaultStore || null;
};

export const getPincodeFromLatLng = async ({ lat, lng }: {lat: number; lng: number}): Promise<string> => {
  const geocoder = new window.google.maps.Geocoder();
  const pincode = new Promise<string>((resolve, reject) => {
    geocoder.geocode({ location: new window.google.maps.LatLng(lat, lng) }, (results, status) => {
      if (status === window.google.maps.GeocoderStatus.OK) {
        const len = results[0].address_components.length;
        resolve(results[0].address_components[len - 1].long_name);
      } else {
        reject();
      }
    });
  });
  return pincode;
};
