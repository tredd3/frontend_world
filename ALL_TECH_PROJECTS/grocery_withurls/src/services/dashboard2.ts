import promiseMemoize from 'promise-memoize';
import { APIDashboardProduct } from '../types/api';
import makeApiCall from './helpers/make-api-call';
import { apiRoutes } from './api-routes';
import { DashboardProductPivot } from '../types';
import { pivotDashboardProduct } from './helpers/pivots';
import { getUser } from './user';

type Banner = { id: number; imageUrl: string };
type BannerWithTitle = Banner & { title: string };
type WidgetOfType<T> = { type: T; subType: number };
type BannerWidget<T> = WidgetOfType<T> & { banners: Banner[] };

export type HorizontalScrollableIcons = WidgetOfType<1> & { banners: BannerWithTitle[] };
export type Pager = BannerWidget<3>;
export type ThinCarouselBanner = BannerWidget<4>;
export type BigSquareCarouselBanner = BannerWidget<5>;
export type SmallSquareCarouselBanner = BannerWidget<6> & { title: string }
export type HorizontalProducts = WidgetOfType<7> & {
  title: string;
  viewAllLabel: string;
  products: DashboardProductPivot[];
};
export type W4x = WidgetOfType<11> & { banners: BannerWithTitle[] };
export type W9x = WidgetOfType<12> & { banners: BannerWithTitle[] } & { title: string };
export type W4xWithBanner = BannerWidget<15> & { title: string };
export type W9xWithBanner = BannerWidget<16> & { title: string };

export type Widget =
  HorizontalScrollableIcons
  | Pager
  | ThinCarouselBanner
  | BigSquareCarouselBanner
  | SmallSquareCarouselBanner
  | HorizontalProducts
  | W4x
  | W9x
  | W4xWithBanner
  | W9xWithBanner

type DashboardRequestType = {
  pincode?: number;
  storeId?: number;
};

type APIBanner = {
  bannerId: number;
  imageUrl: string;
};

type APIMiscellaneous = {
  name: string;
  mscId: number;
  imageUrl: string;
}

type APIWidget = {
  type: Widget['type'];
  subType: number;
  title: string;
  deepLinkText: string;
  data: {
    banners: null | APIBanner[];
    products: null | APIDashboardProduct[];
    miscellaneous: null | APIMiscellaneous[];
  };
};

export type DashboardResponseType = {
  widgets: APIWidget[];
};

const widgetBase = (widget: APIWidget) => ({
  type: widget.type,
  subType: widget.subType
});

const apiBannerToBanner = (banner: APIBanner | APIMiscellaneous): Banner => ({
  id: (banner as APIBanner).bannerId || (banner as APIMiscellaneous).mscId,
  imageUrl: banner.imageUrl
});

const createBanners = ({ data: widgetData }: APIWidget): { banners: Banner[] } => ({
  banners: ((widgetData.banners || widgetData.miscellaneous || []) as (APIBanner | APIMiscellaneous)[])
    .map(apiBannerToBanner)
});

const pivotBanner = (widget: APIWidget) => ({
  ...widgetBase(widget),
  ...createBanners(widget)
});

const createBannersWithTitle = (widget: APIWidget): { banners: BannerWithTitle[] } => ({
  banners: ((widget.data.miscellaneous || [])as APIMiscellaneous[])
    .map(banner => ({
      ...apiBannerToBanner(banner),
      title: banner.name
    }))
});

const createProducts = (widget: APIWidget): { products: DashboardProductPivot[] } => ({
  products: widget.data.products!.map(pivotDashboardProduct)
});

const pivotBannersWithTitle = (widget: APIWidget) => ({
  ...widgetBase(widget),
  ...createBannersWithTitle(widget)
});

const pivotProducts = (widget: APIWidget) => ({
  ...widgetBase(widget),
  ...createProducts(widget)
});

export const pivotDashboardResponse = ({ widgets }: DashboardResponseType) => (
  widgets
    .filter(w => w.data.banners || w.data.miscellaneous || w.data.products)
    .map(widget => {
      switch (widget.type) {
        case 1: return pivotBannersWithTitle(widget) as HorizontalScrollableIcons;
        case 3: return pivotBanner(widget) as Pager;
        case 4: return pivotBanner(widget) as ThinCarouselBanner;
        case 5: return pivotBanner(widget) as BigSquareCarouselBanner;
        case 6:
          return {
            ...widgetBase(widget),
            title: widget.title,
            ...createBanners(widget)
          } as SmallSquareCarouselBanner;
        case 7:
          return {
            title: widget.title,
            viewAllLabel: widget.deepLinkText,
            ...pivotProducts(widget)
          } as HorizontalProducts;
        case 11: return pivotBannersWithTitle(widget) as W4x;
        case 12:
          return {
            title: widget.title,
            ...pivotBannersWithTitle(widget)
          } as W9x;
        case 15: return {
          ...widgetBase(widget),
          title: widget.title,
          ...createBanners(widget)
        } as W4xWithBanner;
        case 16: return {
          ...widgetBase(widget),
          title: widget.title,
          ...createBanners(widget)
        } as W9xWithBanner;
        default: return null;
      }
    })
    .filter(<T>(x: T | null): x is T => x !== null)
);

export const getWidgetDefinitions = promiseMemoize(async () => {
  const { preferences: { pincode, storeId = 0 } } = await getUser();

  const apiWidgets = await makeApiCall<DashboardRequestType, DashboardResponseType>(apiRoutes.getDashboard, {
    body: { pincode, storeId }
  });

  return pivotDashboardResponse(apiWidgets);
}, { resolve: 'json' });
