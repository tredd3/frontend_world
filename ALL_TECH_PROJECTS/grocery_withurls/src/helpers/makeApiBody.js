import store from '../store';
import { getUser } from '../services/user';
import { getCart } from '../services/cart';

const getCartId = () => store.getState().cart.CartId || 0;
const getSelectedStoreId = async () => {
  const user = await getUser();
  const { preferences: { storeId } } = user;
  return storeId || undefined;
};

export const getUserId = () => store.getState().user.userId;

export const addCartApiBody = async ({
  skuId, Quantity = 1, ProductType = 0, deliveryModes, addressIdForCheckout
}) => {
  let apiBody = {};

  apiBody = {
    product: [{
      skuId, qty: Number(Quantity), productType: ProductType
    }]
  };

  // Here action is mandatory parameter to be passed. Because based on it backend internally calls checkout api.
  // action : 1 means checkout call
  // action : 0 means no checkout call
  // Also we need to pass other parameters like deliveryDate, timeSlot which needed for checkout api call at backend
  // when action is 1

  if (deliveryModes && Object.keys(deliveryModes).length !== 0) {
    const { date, timeSlot } = deliveryModes;
    apiBody = {
      ...apiBody, action: 1, deliveryDate: date, timeslot: timeSlot, customerAddressId: addressIdForCheckout
    };
  } else {
    apiBody = { ...apiBody, action: 0 };
  }

  const user = await getUser();
  if (user) {
    const { userId } = user;
    apiBody = { ...apiBody, userId };
  }

  const storeId = await getSelectedStoreId();
  if (storeId) {
    apiBody = { ...apiBody, storeId };
  }
  try {
    const cart = await (storeId ? getCart(storeId) : getCart());
    if (cart && cart.cartId !== 0) {
      const { cartId } = cart;
      apiBody = { ...apiBody, cartId };
    }
  } catch (e) {
    console.log(e.message);
  }

  return apiBody;
};

export const removeCartApiBody = ({ skuId, Quantity = 0, ProductType = 0 }) => ({
  cartId: getCartId(),
  skuId,
  qty: Quantity,
  type: ProductType
});

export const moveItemApiBody = ({
  skuId, Quantity = 0, ProductType = 0, moveType
}) => ({
  cartId: getCartId(),
  skuId,
  qty: Quantity,
  productType: ProductType,
  moveType
});

export const removeSaveLaterCartApiBody = ({
  skuId, Quantity = 0, ProductType = 0, moveType
}) => ({
  skuId,
  qty: Quantity,
  productType: ProductType,
  productCategoryType: moveType
});
