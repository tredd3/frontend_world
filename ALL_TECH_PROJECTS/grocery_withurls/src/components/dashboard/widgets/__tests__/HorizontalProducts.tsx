import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { mocked } from 'ts-jest/utils';
import HorizontalProductsWidget from '../HorizontalProducts';
import { pivotDashboardResponse, HorizontalProducts, DashboardResponseType } from '../../../../services/dashboard2';
import mockDB from '../../../../mocks/db.json';
import * as analytics from '../../../../helpers/analytics';

jest.mock('../../../../helpers/analytics');
jest.mock('../../../uiControls/horizontalProductsList/HorizontalProductsList', () => {
  const MockedHorizontalProductsList: React.FC = ({ children }) => (
    <div>
      <div>HorizontalProductsList placeholder</div>
      {children}
    </div>
  );

  return {
    __esModule: true,
    default: MockedHorizontalProductsList
  };
});

const { setPurchaseJourney, trackLink } = mocked(analytics);

const mockWidgetData = pivotDashboardResponse(mockDB.dashboard.Data as DashboardResponseType)
  .find(({ type }) => type === 7) as HorizontalProducts;

describe('HorizontalProductsWidget tests', () => {
  beforeEach(() => {
    setPurchaseJourney.mockClear();
    trackLink.mockClear();
  });

  it('should render the widget correctly', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <HorizontalProductsWidget
          widgetData={mockWidgetData}
        />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should handle click action for view all link', () => {
    const { getByText } = render(
      <MemoryRouter>
        <HorizontalProductsWidget
          widgetData={mockWidgetData}
        />
      </MemoryRouter>
    );

    fireEvent.click(getByText(mockWidgetData.viewAllLabel));
    expect(setPurchaseJourney).toHaveBeenCalledTimes(1);
    expect(setPurchaseJourney.mock.calls[0]).toMatchSnapshot();
    expect(trackLink).toHaveBeenCalledTimes(1);
    expect(trackLink.mock.calls[0]).toMatchSnapshot();
  });
});
