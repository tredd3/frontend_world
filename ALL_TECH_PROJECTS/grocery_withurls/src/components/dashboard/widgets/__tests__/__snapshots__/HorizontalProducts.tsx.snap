// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HorizontalProductsWidget tests should handle click action for view all link 1`] = `
Array [
  "Deals",
]
`;

exports[`HorizontalProductsWidget tests should handle click action for view all link 2`] = `
Array [
  "See All",
  "Todays Deals",
  "Middle",
]
`;

exports[`HorizontalProductsWidget tests should render the widget correctly 1`] = `
<DocumentFragment>
  .emotion-3 {
  padding: 14px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-bottom: 10px;
}

.emotion-0 {
  color: #000;
  font-weight: 600;
  font-size: 16px;
}

.emotion-1 {
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: none;
  color: #0066C0;
  font-size: 14px;
}

.emotion-1:hover,
.emotion-1:focus,
.emotion-1:visited,
.emotion-1:active {
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: none;
}

<section
    class="emotion-3"
  >
    <div
      class="emotion-2"
    >
      <span
        class="emotion-0"
      >
        Hot Deals
      </span>
      <a
        class="emotion-1"
        href="/collections/-1?type=7&subtype=6"
      >
        View All
      </a>
    </div>
    <div>
      <div>
        HorizontalProductsList placeholder
      </div>
    </div>
  </section>
</DocumentFragment>
`;
